ICU-20520 UMutex refactoring.

aheninger opened the pull request:
- Use STATIC_NEW for mutex creation, to avoid order-of-destruction problems
    by avoiding destruction altogether, while avoiding memory leak reports.

  - Remove UConditionVar, replace with direct use of std::condition_variable

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20520
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

aheninger commented:
Another go at fixing our mutexes.

For reviewing, Peter, does this stop the crashes you have been seeing?

Jeff, the file common/cmemory.h has a Microsoft conditional. Does it look safe?
I found a web page a while back suggesting that the compiler behavior was a bug that was being addressed, but didn't bookmark it and can't find it again.

jefgen commented:
Thanks Andy.

Based on the page here: https://devblogs.microsoft.com/oldnewthing/20150220-00/?p=44623
I think the MSVC conditional would still be required as it appears that the MSVC compiler will actually generate functions for all three calling conventions -- but they will be discarded at link time.

I wonder if maybe using a `static_cast` might help make the type more clear?

```c++
#ifdef _MSC_VER
#define STATIC_NEW(type) (* static_cast<type *(__cdecl *)(void)>( [] () { \
    alignas(type) static char storage[sizeof(type)]; \
    return new(storage) type;}))();
#else
#define STATIC_NEW(type) (*[] () { \
    alignas(type) static char storage[sizeof(type)]; \
    return new(storage) type;})();
#endif
```

Though I suppose "clear" it is somewhat debatable, with all the various brackets/parens...

pedberg-icu commented:
> For reviewing, Peter, does this stop the crashes you have been seeing?

I just rolled this in to Apple ICU and will be sending out versions for testing, so I may not know for a day or two. I had not experienced the crashes myself so I will need to get feedback fro people who had.

pedberg-icu commented on icu4c/source/common/locid.cpp:
```diff
@@ -38,6 +38,7 @@
 #include "unicode/strenum.h"
 #include "unicode/stringpiece.h"
 #include "unicode/uloc.h"
+#include "cmemory.h"
```
This is redundant, locid.cpp already includes cmemory.h about 5 lines after this

pedberg-icu commented on icu4c/source/i18n/numfmt.cpp:
```diff
@@ -1362,8 +1362,8 @@ NumberFormat::makeInstance(const Locale& desiredLocale,
-        static icu::UMutex nscacheMutex = U_MUTEX_INITIALIZER;
-        Mutex lock(&nscacheMutex);
+        static icu::UMutex *nscacheMutex = STATIC_NEW(UMutex);
+        Mutex lock(nscacheMutex);
```
Should we use STATIC_NEW(icu::UMutex) for consistency in namespace usage?

pedberg-icu commented on icu4c/source/i18n/reldatefmt.cpp:
```diff
@@ -1184,8 +1184,8 @@ UnicodeString& RelativeDateTimeFormatter::adjustForContext(UnicodeString &str) c
-    static icu::UMutex gBrkIterMutex = U_MUTEX_INITIALIZER;
-    Mutex lock(&gBrkIterMutex);
+    static icu::UMutex *gBrkIterMutex = STATIC_NEW(UMutex);
+    Mutex lock(gBrkIterMutex);
```
Should we use STATIC_NEW(icu::UMutex) for consistency in namespace usage?

pedberg-icu commented on icu4c/source/common/cmemory.h:
```diff
@@ -97,6 +97,33 @@ typedef union {
+ *   - We don't want it to ever be destructed, to avoid order-of-destruction
+ *     or use-after-destruction problems.
+ *   - We want to avoid an ordianary heap allocated object, to avoid triggering
+ *     memory leak reports, from valgrind, for example.
```
nit: ordianary -> ordinary

aheninger commented:
Incorporated the review comments. Thanks.

For the inconsistent use the icu:: namespace qualification that Peter pointed out, I removed the existing namespace qualifications, rather than adding new ones. It's inside the namespace, and nothing else nearby is qualified.

pedberg-icu commented:
> The documentation in common/mutex.h needs updating (e.g. example at line 38 using U_MUTEX_INITIALIZER)

This still needs fixing

axelandrejs commented on icu4c/source/common/umutex.cpp:
```diff
@@ -99,13 +73,13 @@ umtx_condSignal(UConditionVar *cond) {
-    static std::mutex m;
-    return m;
+    static std::mutex *m = STATIC_NEW(std::mutex);
+    return *m;
```
What happens when the ICU binary gets unloaded? Does this content leak? And if it does, what exactly leaks? Is std::mutex backed by an OS resource such as a critical section that then persists for the lifetime of the process? If so, that needs to be avoided.

jefgen replied:
I think the implementation details of std::mutex might be compiler and/or library dependent.
If it is backed by a Critical Section, then we would need some way to invoke the destructor for this object.

axelandrejs replied:
I looked it up, and the Windows inbox version of the CRT uses critical sections. That means if you're constructing this object you need to make sure it is destructed on DLL exit.

In general I would recommend considering forking the behavior here. There are scenarios where init/uninit is possible in DLL attach/detach is possible and desirable. Those should take advantage of that, and avoid the overhead and correctness issues that come with dynamic usage.

---
In reply to: [279541894](https://github.com/unicode-org/icu/pull/620#discussion_r279541894) [](ancestors = 279541894)

aheninger commented:
Yes, if the ICU library is unloaded, the static mutexes will leak at the OS level.
To the best of my understanding, there is no good way to avoid this.

If static object destructors are allowed to run (local objects, instead of local pointers initialized with new), you can get into order of destruction, use-after-deletion issues. Letting static/global C++ objects leak is a commonly recommended work around for this.

axelandrejs commented:
> Yes, if the ICU library is unloaded, the static mutexes will leak at the OS level.
> To the best of my understanding, there is no good way to avoid this.
> If static object destructors are allowed to run (local objects, instead of local pointers initialized with new), you can get into order of destruction, use-after-deletion issues. Letting static/global C++ objects leak is a commonly recommended work around for this.

Leaking these means that in long-running processes you'll have an ever-increasing leak of system resources, which will eventually catch up with you. As such, unload leaks need to be avoided. This is going to cause us significant issues otherwise. In fact the old 63 implementation already does cause us issues - that's how this PR caught my eye:)

You're right, relying on C++ cleanup can cause issues. Techniques to get around this include

1. Ref counting the global objects to caller-initialized objects. That is, when the last caller object is closed the globals go away. This can be made safe with interlocked operations.

2. Do the uninit in DLL detach. Clean up the global mutex as the last thing. This is safe because if anything is going to use something in ICU after that point it's going to crash anyway as the executable code is going away right after. So any such usage would be a bug with or without this.

We're using these techniques for various system binaries, and we have neither crashes nor unload leaks, so this is definitely doable. As I mentioned earlier, I think the best approach would be to have a "use static init" define, which would make the object global and do init/uninit in dllmain (or other platform-appropriate place). One could also imagine a hybrid approach of doing the init on-demand and the uninit in DLL detach, but that might be overkill.

jefgen commented:
I found a section here https://isocpp.org/wiki/faq/ctors#nifty-counter-idiom on the ISO C++ FAQ that suggests using a technique named the "Nifty Counter Idiom" for the use case where "you need control the order of both static initialization and static deinitialization". But it had a giant "TODO: Write this up" and didn't give any details on it.
However, I found an example here https://en.wikibooks.org/wiki/More_C++_Idioms/Nifty_Counter  of the idiom. It looks somewhat similar to the approach suggested by Axel of "Ref counting the global objects to caller-initialized objects". But it also uses placement new into static storage in addition to a ref count.
(As an aside, I suppose it makes sense that the ISO C++ FAQ would suggest using that instead of uninit in DLL detach, since DLL detach would be inherently platform specific).

Maybe a similar technique to the "Nifty Counter Idiom" could be used? Since the UMutex is internal to the library and there is C++11 now, the ref counter could perhaps use std::atomic maybe?

aheninger commented:
Hmm,

Cleaning up the OS level mutexes would be something new for ICU; in all previous versions, mutexes have been allowed to leak; the OS recovers them when a process terminates.

The whole leakage question only comes into play if an application/process is repeatedly DLL unloading and reloading ICU. In the normal case - once loaded, ICU stays loaded until the process terminates - there is a fixed set of static mutexes; the number of them does not grow.

ICU already has a generic mechanism for cleaning up, [u_cleanup(),](https://github.com/unicode-org/icu/blob/master/icu4c/source/common/unicode/uclean.h), and it already takes into account dependency ordering between the various ICU services. If we really want to clean up after mutexes, perhaps we could add it there.

In any event, I think we should separate the cleanup question from this PR and issue. We need the changes here to dodge the use-after-delete order-of-destruction issues that Peter saw.

axelandrejs commented:
> Hmm,
> Cleaning up the OS level mutexes would be something new for ICU; in all previous versions, mutexes have been allowed to leak; the OS recovers them when a process terminates.
> The whole leakage question only comes into play if an application/process is repeatedly DLL unloading and reloading ICU. In the normal case - once loaded, ICU stays loaded until the process terminates - there is a fixed set of static mutexes; the number of them does not grow.
> ICU already has a generic mechanism for cleaning up, u_cleanup(),, and it already takes into account dependency ordering between the various ICU services. If we really want to clean up after mutexes, perhaps we could add it there.
> In any event, I think we should separate the cleanup question from this PR and issue. We need the changes here to dodge the use-after-delete order-of-destruction issues that Peter saw.

Unfortunately those issues cannot be fully separated. Your change fixes the order-of-destruction issue but in the process introduces new leaks. If this goes in as-is then there needs to be a 65 release blocking ticket to fix it as I am not sure we could ship this as-is. Unload of the DLL is a real, and pretty frequent, scenario.

I do like the approach of adding it to u_cleanup. That way we or anyone else who cares can call that during detach. I still think that we should have a version of the code that does just that during detach so that we don't ask random consumers to modify ICU code. But that is a less critical concerm.

aheninger commented:
So here is where we stand:

ICU 63, 62, 61, ... back for many years:  Mutexes leak on DLL unload. No order of destruction problems.

ICU 64:  Mutexes don't leak, but have order of destruction problems that may affect applications that have static instances of ICU types.

ICU 65: The order of destruction problems must be fixed. Beyond that, either
  a. Go back to ICU 63 and earlier behavior, with leaking mutexes.
  b. Introduce something new to stop the leaks.

If we choose (b), it needs to be discussed. It would be a non-trivial change. The non-leak behavior of 64 was somewhat accidental, it wasn't planned or advertised. Choosing (b) does not preclude doing (a) first, to fix the regression from ICU 63.

This PR is intended to do (a). Barring problems other than the leaks, I would like to get it in.

axelandrejs commented:
That is fair. Please file a ticket tracking because as I said, we will need a solution for b) before 65 ships. Going back to a) is a regression for us.

(this was less of a problem in prior releases because usage of ICU was more limited; now that we use it more this is likely a ship blocker for us)

axelandrejs commented on icu4c/source/common/umutex.cpp:
```diff
@@ -43,8 +43,8 @@ U_NAMESPACE_BEGIN
 
 // The ICU global mutex. Used when ICU implementation code passes NULL for the mutex pointer.
 static UMutex *globalMutex() {
```
To facilitate future cleanup efforts, please move the definition of this and all other mutexes using the same mechanism to headers.

axelandrejs replied:
To avoid the ordering issue, you'll probably need to pull all the mutex declarations in one place so that you can then destroy them in a predictable manner on unload. But yes that can wait until that problem is addressed.

---
In reply to: [279954961](https://github.com/unicode-org/icu/pull/620#discussion_r279954961) [](ancestors = 279954961)

aheninger replied:
> To avoid the ordering issue, you'll probably need to pull all the mutex declarations in one place ...

ICU already has a clean-up mechanism. It does not work like you are thinking.

Each ICU module/service/file/whatever that needs cleanup registers a cleanup function with ICU's clean up service. The registration includes an ordering, to avoid dependency problems.

Exactly what each module's cleanup function needs to handle is not surfaced globally; that is purely the internal business of that module. Exporting that kind of internal detail to more global scopes would not be desirable.

Anything unloading the ICU DLLs already must use u_cleanup(), to avoid memory leaks from ICU's internal caches. Many ICU modules using mutexes already have cleanup functions in place already that could be extended.

Note also, ICU must continue to operate correctly if use resumes after calling u_cleanup() (without unloading and reloading the library). This means that function local static init cannot be used as a lazy init mechanism for anything that is deleted by u_cleanup().

In any event, this PR is not the right place to design mutex cleanup. It's just to get us back to ICU 63 functionality, which is currently broken and causing immediate problems.

aheninger commented:
New issue for closing mutexes as part of u_cleanup() 
https://unicode-org.atlassian.net/browse/ICU-20588

pedberg-icu commented:
> > For reviewing, Peter, does this stop the crashes you have been seeing?
> 
> I just rolled this in to our version of ICU and will be sending out versions for testing, so I may not know for a day or two. I had not experienced the crashes myself so I will need to get feedback from people who had.

I have had Apple ICU versions with this fix available for limited testing for 5 days and more general testing for 3 days, and have not yet seen any reports of the mutex crash with these versions.
